Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,197 @@
+            guard let port = UInt16(from), port > 0 else {
+                throw MongoError.InvalidArgumentError(
+                    message: "port must be a valid, positive unsigned 16 bit integer"
+                )
+            }
+            return port
+        }
+
+        private enum HostType: String {
+            case ipv4
+            case ipLiteral = "ip_literal"
+            case hostname
+            case unixDomainSocket
+        }
+
+        /// The hostname or IP address.
+        public let host: String
+
+        /// The port number.
+        public let port: UInt16?
+
+        private let type: HostType
+
+        /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
+        internal init(_ hostAndPort: String = "localhost:27017", scheme: Scheme) throws {
+            guard !hostAndPort.contains("?") else {
+                throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
+            }
+
+            var host: String
+            var port: UInt16?
+
+            // Check if host is an IPv6 literal.
+            if hostAndPort.first == "[" {
+                let ipLiteralRegex = try NSRegularExpression(pattern: #"^\[(.*)\](?::([0-9]+))?$"#)
+                guard
+                    let match = ipLiteralRegex.firstMatch(
+                        in: hostAndPort,
+                        range: NSRange(hostAndPort.startIndex..<hostAndPort.endIndex, in: hostAndPort)
+                    ),
+                    let hostRange = Range(match.range(at: 1), in: hostAndPort)
+                else {
+                    throw MongoError.InvalidArgumentError(message: "couldn't parse address from \(hostAndPort)")
+                }
+                host = String(hostAndPort[hostRange])
+                if let portRange = Range(match.range(at: 2), in: hostAndPort) {
+                    port = try HostIdentifier.verifyPort(from: String(hostAndPort[portRange]), scheme: scheme)
+                } else {
+                    port = scheme == .srv ? nil : 27017
+                }
+                self.type = .ipLiteral
+            } else {
+                let parts = hostAndPort.components(separatedBy: ":")
+                host = String(parts[0])
+                guard parts.count <= 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "expected only a single port delimiter ':' in \(hostAndPort)"
+                    )
+                }
+                if parts.count > 1 {
+                    port = try HostIdentifier.verifyPort(from: parts[1], scheme: scheme)
+                } else {
+                    port = scheme == .srv ? nil : 27017
+                }
+                self.type = .hostname
+            }
+            self.host = host
+            self.port = port
+        }
+
+        public var description: String {
+            switch self.type {
+            case .ipLiteral:
```

Comment:
This could be condensed a little bit to have the `switch` construct the host portion, and then at the end have a single conditional that handles appending the port.

```swift
// because host will be initialized in all branches of the switch, we can leave it uninitialized here
// the Swift compiler will complain if there is a branch that doesn't initialize it
let host: String
switch self.type {
// ...
}
if let port = self.port {
    // ..
}
// ...
```

Reply:
done